src/naive/datetime/mod.rs

```diff
@@ -131,6 +131,80 @@ impl NaiveDateTime {
         datetime.expect("invalid or out-of-range datetime")
     }
 
+    /// Makes a new `NaiveDateTime` corresponding to a UTC date and time,
+    /// from the number of non-leap milliseconds
+    /// since the midnight UTC on January 1, 1970 (aka "UNIX timestamp")
+    ///
+    /// Returns `None` on the out-of-range number of milliseconds.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use chrono::{NaiveDateTime, NaiveDate};
+    ///
+    ///
+    /// let dt = NaiveDate::from_ymd(1970, 1, 1).and_hms_milli(0, 0, 0, 42);
+    ///
+    /// let got = NaiveDateTime::from_timestamp_millis(dt.timestamp_millis()).unwrap();
+    /// assert_eq!(dt, got);
+    /// ```
+    #[inline]
+    pub fn from_timestamp_millis(milliseconds: i64) -> Option<NaiveDateTime> {
+        let timestamp_seconds = milliseconds / 1000;
+        NaiveDateTime::from_timestamp_opt(timestamp_seconds, (milliseconds % 1000 * 1000000) as u32)
```

Comment:
I think the conversion to nanoseconds is wrong here. `milliseconds % 1000 * 1000000` might still be negative and the cast to `u32` would overflow. Same for the other conversions.